src/book/02-projects/01-nubots/03-build-system.mdx

```diff
@@ -5,3 +5,265 @@ title: Build System
+
+Two common extensions are `Configuration` and `Script`. `Configuration` allows modules to monitor their configuration file(s) and apply updates in real-time. `Script` provides similar functionality, but for script files.
+
+#### Messages
+
+Every message in the NUbots codebase is a [protobuf](https://developers.google.com/protocol-buffers) message and all protobuf messages live in the `shared/message` folder.
+
+The general convention is to groups message files into folders based on the meta-module which introduces the message to the system. For instance, `module/input/Camera` introduces `Image` messages to the system, so `Image.proto` lives in `shared/message/input`.
+
+NUClear Roles provides special wrappers for certain types. Vector and matrix types are mapped to [Eigen](http://eigen.tuxfamily.org/index.php?title=Main_Page) types. To use these mappings simply use `import Vector.proto;` and `import Matrix.proto;`. Some examples of types that you can use are:
+
+- `fvec3`: A vector with 3 components, maps to an `Eigen::Vector3f`
+- `vec3`: A vector with 3 components, maps to an `Eigen::Vector3d`
+- `mat3`: A 3x3 matrix, maps to an `Eigen::Matrix3d`
+
+All messages are packaged in a similar fashion to the C++ namespaces used for modules. For instance, `Image.proto` is in the `message.input` package.
+
+NUClear Roles creates a C++ wrapper class for each message to create a simpler interface for accessing and mutating message fields. The C++ namespace for a message follows its package. For instance, the `message.input.Image` message will have a C++ namespace of `message::input::Image`.
+
+#### Utilities
+
+Utilities are collections of functions that perform useful tasks. Typically, these functions are stateless and are (or feasibly could be) used by multiple different modules (or even other utilities).
+
+As is the case with modules and messages, utilities are grouped together into meta-utilities. For instance, utilities dealing with angles, coordinate system transforms, and geometrical constructs live in the `shared/utility/math` meta-utility.
+
+#### Tools and the `b` command
+
+NUClear Roles introduces the `b` command that wraps up all of the functionality needed to create modules and build the NUbots code.
+
+| Command   | Description                                       |
+| --------- | ------------------------------------------------- |
+| configure | Configure the CMake system for the current target |
+| build     | Build the currently configured code               |
+| edit      | Edit configuration files                          |
+| format    | Applies code formatting rules to codebase         |
+| install   | Install build code to specified robot             |
+| module    | Generate new modules                              |
+| run       | Run a role locally                                |
+| target    | Select a target to work on                        |
```

Comment:
Missing some important ones here, e.g. `nbs` and some less important ones like `foot_down`